Mypy is pretty upset with the QPROGRAM type

Not sure what the deal is quite yet, but running `mypy` (the static type checker) gives lots of errors of this form. Might be worth investigating?

```
mitiq/mitiq_pyquil/tests/test_zne.py:21: error: Variable "mitiq.QPROGRAM" is not valid as a type
mitiq/mitiq_pyquil/tests/test_zne.py:21: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
```

--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@ all: dist
 build: check-all docs test-all
 
 .PHONY: check-all
-check-all: check-format check-style
+check-all: check-format check-style check-types
 
 .PHONY: check-format
 check-format:
@@ -14,6 +14,10 @@ check-format:
 .PHONY: check-style
 check-style:
 	flake8
+
+.PHONY: check-types
+check-types:
+	mypy .
 
 .PHONY: clean
 clean:

--- a/mitiq/_about.py
+++ b/mitiq/_about.py
@@ -27,7 +27,10 @@ from scipy import __version__ as scipy_version
 
 import mitiq
 
-MITIQ_INSTALL_PATH = os.path.dirname(inspect.getsourcefile(mitiq))
+MITIQ_SOURCE_FILE = inspect.getsourcefile(mitiq)
+assert MITIQ_SOURCE_FILE
+
+MITIQ_INSTALL_PATH = os.path.dirname(MITIQ_SOURCE_FILE)
 PYTHON_VERSION = sys.version_info[0:3]
 
 

--- a/mitiq/_typing.py
+++ b/mitiq/_typing.py
@@ -28,19 +28,19 @@ which is then handled gracefully by the Union type.
 """
 from typing import Union
 
-from cirq import Circuit
+from cirq import Circuit as _Circuit
 
 try:
-    from pyquil import Program
-except ImportError:
-    Program = Circuit
+    from pyquil import Program as _Program
+except ImportError:  # pragma: no coverage
+    _Program = _Circuit
 
 try:
-    from qiskit import QuantumCircuit
-except ImportError:
-    QuantumCircuit = Circuit
+    from qiskit import QuantumCircuit as _QuantumCircuit
+except ImportError:  # pragma: no coverage
+    _QuantumCircuit = _Circuit
 
-QPROGRAM = Union[Circuit, Program, QuantumCircuit]
+QPROGRAM = Union[_Circuit, _Program, _QuantumCircuit]
 
 SUPPORTED_PROGRAM_TYPES = {
     "cirq": "Circuit",

--- a/mitiq/_typing.py
+++ b/mitiq/_typing.py
@@ -15,27 +15,35 @@
 
 """This file is used to optionally import what program types should be allowed
 by mitiq based on what packages are installed in the environment.
+
+Supported program types:
+    cirq.Circuit
+    pyquil.Program
+    qiskit.QuantumCircuit
+
+When pyquil or qiskit are not available for import, we take advantage of a
+clever hack in the except ImportError block, setting the type alias
+corresponding to the class that we failed to import equal to cirq.Circuit,
+which is then handled gracefully by the Union type.
 """
 from typing import Union
+
+from cirq import Circuit
+
+try:
+    from pyquil import Program
+except ImportError:
+    Program = Circuit
+
+try:
+    from qiskit import QuantumCircuit
+except ImportError:
+    QuantumCircuit = Circuit
+
+QPROGRAM = Union[Circuit, Program, QuantumCircuit]
 
 SUPPORTED_PROGRAM_TYPES = {
     "cirq": "Circuit",
     "pyquil": "Program",
     "qiskit": "QuantumCircuit",
 }
-
-AVAILABLE_PROGRAM_TYPES = {}
-
-for (module, program_type) in SUPPORTED_PROGRAM_TYPES.items():
-    try:
-        exec(f"from {module} import {program_type}")
-        AVAILABLE_PROGRAM_TYPES.update({module: program_type})
-    except ImportError:
-        pass
-
-QPROGRAM = Union[
-    tuple(
-        f"{package}.{circuit}"
-        for package, circuit in AVAILABLE_PROGRAM_TYPES.items()
-    )
-]

--- a/mitiq/benchmarks/maxcut.py
+++ b/mitiq/benchmarks/maxcut.py
@@ -33,7 +33,7 @@ SIMULATOR = DensityMatrixSimulator()
 
 def make_noisy_backend(
     noise: float, obs: np.ndarray
-) -> Callable[[Circuit, int], float]:
+) -> Callable[[Circuit], float]:
     """ Helper function to match mitiq's backend type signature.
 
     Args:
@@ -44,7 +44,7 @@ def make_noisy_backend(
         A mitiq backend function.
     """
 
-    def noisy_backend(circ: Circuit):
+    def noisy_backend(circ: Circuit) -> float:
         return noisy_simulation(circ, noise, obs)
 
     return noisy_backend
@@ -76,7 +76,7 @@ def make_maxcut(
     """
     # get the list of unique nodes from the list of edges
     nodes = list({node for edge in graph for node in edge})
-    nodes = range(max(nodes) + 1)
+    nodes = list(range(max(nodes) + 1))
 
     # one qubit per node
     qreg = [NamedQubit(str(nn)) for nn in nodes]

--- a/mitiq/benchmarks/random_circuits.py
+++ b/mitiq/benchmarks/random_circuits.py
@@ -23,6 +23,7 @@ from cirq import NamedQubit, Circuit
 from mitiq import execute_with_zne
 from mitiq._typing import QPROGRAM
 from mitiq.zne.inference import Factory
+from mitiq.zne.scaling import fold_gates_at_random
 from mitiq.benchmarks.utils import noisy_simulation
 
 
@@ -61,7 +62,7 @@ def rand_circuit_zne(
     trials: int,
     noise: float,
     fac: Factory = None,
-    scale_noise: Callable[[QPROGRAM, float], QPROGRAM] = None,
+    scale_noise: Callable[[QPROGRAM, float], QPROGRAM] = fold_gates_at_random,
     op_density: float = 0.99,
     silent: bool = True,
     seed: Optional[int] = None,

--- a/mitiq/benchmarks/randomized_benchmarking.py
+++ b/mitiq/benchmarks/randomized_benchmarking.py
@@ -54,17 +54,26 @@ def rb_circuits(
         qubit1 = NamedQubit("0")
         if n_qubits == 1:
             rb_circuits = [
-                _random_single_q_clifford(qubit1, num, C1, CFD_MAT_1Q)  # type: ignore
+                _random_single_q_clifford(
+                    qubit1,  # type: ignore
+                    num,
+                    C1,
+                    CFD_MAT_1Q,
+                )
                 for _ in range(trials)
             ]
         elif n_qubits == 2:
             qubit2 = NamedQubit("1")
             cfd_matrices = _two_qubit_clifford_matrices(
-                qubit1, qubit2, CLIFFORDS  # type: ignore
+                qubit1, qubit2, CLIFFORDS,  # type: ignore
             )
             rb_circuits = [
                 _random_two_q_clifford(
-                    qubit1, qubit2, num, cfd_matrices, CLIFFORDS  # type: ignore
+                    qubit1,  # type: ignore
+                    qubit2,  # type: ignore
+                    num,
+                    cfd_matrices,
+                    CLIFFORDS,
                 )
                 for _ in range(trials)
             ]

--- a/mitiq/benchmarks/randomized_benchmarking.py
+++ b/mitiq/benchmarks/randomized_benchmarking.py
@@ -54,17 +54,17 @@ def rb_circuits(
         qubit1 = NamedQubit("0")
         if n_qubits == 1:
             rb_circuits = [
-                _random_single_q_clifford(qubit1, num, C1, CFD_MAT_1Q)
+                _random_single_q_clifford(qubit1, num, C1, CFD_MAT_1Q)  # type: ignore
                 for _ in range(trials)
             ]
         elif n_qubits == 2:
             qubit2 = NamedQubit("1")
             cfd_matrices = _two_qubit_clifford_matrices(
-                qubit1, qubit2, CLIFFORDS
+                qubit1, qubit2, CLIFFORDS  # type: ignore
             )
             rb_circuits = [
                 _random_two_q_clifford(
-                    qubit1, qubit2, num, cfd_matrices, CLIFFORDS
+                    qubit1, qubit2, num, cfd_matrices, CLIFFORDS  # type: ignore
                 )
                 for _ in range(trials)
             ]

--- a/mitiq/benchmarks/utils.py
+++ b/mitiq/benchmarks/utils.py
@@ -14,9 +14,11 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 """Utility functions for benchmarking."""
+from typing import cast
+
 import numpy as np
 
-from cirq import Circuit, depolarize, DensityMatrixSimulator
+from cirq import Circuit, depolarize, DensityMatrixSimulator, DensityMatrixTrialResult
 
 SIMULATOR = DensityMatrixSimulator()
 
@@ -33,7 +35,8 @@ def noisy_simulation(circ: Circuit, noise: float, obs: np.ndarray) -> float:
         The observable's expectation value.
     """
     circuit = circ.with_noise(depolarize(p=noise))
-    rho = SIMULATOR.simulate(circuit).final_density_matrix
+    result = cast(DensityMatrixTrialResult, SIMULATOR.simulate(circuit))
+    rho = result.final_density_matrix
     # measure the expectation by taking the trace of the density matrix
     expectation = np.real(np.trace(rho @ obs))
     return expectation

--- a/mitiq/benchmarks/utils.py
+++ b/mitiq/benchmarks/utils.py
@@ -18,7 +18,12 @@ from typing import cast
 
 import numpy as np
 
-from cirq import Circuit, depolarize, DensityMatrixSimulator, DensityMatrixTrialResult
+from cirq import (
+    Circuit,
+    depolarize,
+    DensityMatrixSimulator,
+    DensityMatrixTrialResult,
+)
 
 SIMULATOR = DensityMatrixSimulator()
 

--- a/mitiq/utils.py
+++ b/mitiq/utils.py
@@ -15,14 +15,15 @@
 
 """Utility functions."""
 from copy import deepcopy
+from typing import cast
 
 import numpy as np
 
-from cirq import Circuit, CircuitDag, Gate, Moment
+from cirq import Circuit, CircuitDag, EigenGate, Gate, GateOperation, Moment
 from cirq.ops.measurement_gate import MeasurementGate
 
 
-def _simplify_gate_exponent(gate: Gate) -> Gate:
+def _simplify_gate_exponent(gate: EigenGate) -> EigenGate:
     """Returns the input gate with a simplified exponent if possible,
     otherwise the input gate is returned without any change.
 
@@ -51,7 +52,11 @@ def _simplify_circuit_exponents(circuit: Circuit) -> None:
         simplified_operations = []
         # Iterate over operations in moment
         for op in moment:
-            simplified_gate = _simplify_gate_exponent(op.gate)
+            assert isinstance(op, GateOperation)
+            if isinstance(op.gate, EigenGate):
+                simplified_gate: Gate = _simplify_gate_exponent(op.gate)
+            else:
+                simplified_gate = op.gate
             simplified_operation = op.with_gate(simplified_gate)
             simplified_operations.append(simplified_operation)
         # Mutate the input circuit
@@ -107,7 +112,8 @@ def _equal(
             circ.batch_remove(measurements)
 
             for i in range(len(measurements)):
-                measurements[i][1].gate.key = ""
+                gate = cast(MeasurementGate, measurements[i][1].gate)
+                gate.key = ""
 
             circ.batch_insert(measurements)
 

--- a/mitiq/zne/inference.py
+++ b/mitiq/zne/inference.py
@@ -812,6 +812,13 @@ class PolyExpFactory(BatchedFactory):
         )
 
 
+# Keep a log of the optimization process storing:
+# noise value(s), expectation value(s), parameters, and zero limit
+OptimizationHistory = List[
+    Tuple[List[Dict[str, float]], List[float], List[float], float]
+]
+
+
 class AdaExpFactory(Factory):
     """Factory object implementing an adaptive zero-noise extrapolation
     algorithm assuming an exponential ansatz y(x) = a + b * exp(-c * x),
@@ -880,11 +887,7 @@ class AdaExpFactory(Factory):
         self.asymptote = asymptote
         self.avoid_log = avoid_log
         self.max_scale_factor = max_scale_factor
-        # Keep a log of the optimization process storing:
-        # noise value(s), expectation value(s), parameters, and zero limit
-        self.history: List[
-            Tuple[List[Dict[str, float]], List[float], List[float], float]
-        ] = []
+        self.history: OptimizationHistory = []
 
     def next(self) -> Dict[str, float]:
         """Returns a dictionary of parameters to execute a circuit at."""

--- a/mitiq/zne/inference.py
+++ b/mitiq/zne/inference.py
@@ -27,11 +27,11 @@ from mitiq import QPROGRAM
 from mitiq.utils import _are_close_dict
 
 
-def _instack_to_scale_factors(instack: dict) -> List[float]:
+def _instack_to_scale_factors(instack: List[Dict[str, float]]) -> List[float]:
     """Extracts a list of scale factors from a list of dictionaries."""
-    if all([isinstance(params, dict) for params in instack]):
-        return [params["scale_factor"] for params in instack]
-    return instack
+    if not all(isinstance(params, dict) for params in instack):
+        raise ValueError("instack must be a list of dictionaries")
+    return [params["scale_factor"] for params in instack]
 
 
 class ExtrapolationError(Exception):
@@ -106,7 +106,7 @@ def mitiq_curve_fit(
             # replace OptimizeWarning with ExtrapolationWarning
             if warn.category is OptimizeWarning:
                 warn.category = ExtrapolationWarning
-                warn.message = _EXTR_WARN
+                warn.message = _EXTR_WARN  # type: ignore
             # re-raise all warnings
             warnings.warn_explicit(
                 warn.message, warn.category, warn.filename, warn.lineno
@@ -144,7 +144,7 @@ def mitiq_polyfit(
         # replace RankWarning with ExtrapolationWarning
         if warn.category is RankWarning:
             warn.category = ExtrapolationWarning
-            warn.message = _EXTR_WARN
+            warn.message = _EXTR_WARN  # type: ignore
         # re-raise all warnings
         warnings.warn_explicit(
             warn.message, warn.category, warn.filename, warn.lineno
@@ -171,9 +171,9 @@ class Factory(ABC):
         particular extrapolation algorithm and can be added to the "__init__"
         method of the associated derived class.
         """
-        self._instack = []
-        self._outstack = []
-        self.opt_params = []
+        self._instack: List[Dict[str, float]] = []
+        self._outstack: List[float] = []
+        self.opt_params: List[float] = []
 
     def push(self, instack_val: dict, outstack_val: float) -> None:
         """Appends "instack_val" to "self._instack" and "outstack_val" to
@@ -196,7 +196,7 @@ class Factory(ABC):
         return np.array(self._outstack)
 
     @abstractmethod
-    def next(self) -> float:
+    def next(self) -> Dict[str, float]:
         """Returns a dictionary of parameters to execute a circuit at."""
         raise NotImplementedError
 
@@ -221,9 +221,7 @@ class Factory(ABC):
         self.opt_params = []
 
     def iterate(
-        self,
-        noise_to_expval: Callable[[float], float],
-        max_iterations: int = 100,
+        self, noise_to_expval: Callable[..., float], max_iterations: int = 100,
     ) -> "Factory":
         """Evaluates a sequence of expectation values until enough
         data is collected (or iterations reach "max_iterations").
@@ -265,7 +263,7 @@ class Factory(ABC):
     def run(
         self,
         qp: QPROGRAM,
-        executor: Callable[[QPROGRAM], float],
+        executor: Callable[..., float],
         scale_noise: Callable[[QPROGRAM, float], QPROGRAM],
         num_to_average: int = 1,
         max_iterations: int = 100,
@@ -884,9 +882,9 @@ class AdaExpFactory(Factory):
         self.max_scale_factor = max_scale_factor
         # Keep a log of the optimization process storing:
         # noise value(s), expectation value(s), parameters, and zero limit
-        self.history = (
-            []
-        )  # type: List[Tuple[List[float], List[float], List[float], float]]
+        self.history: List[
+            Tuple[List[Dict[str, float]], List[float], List[float], float]
+        ] = []
 
     def next(self) -> Dict[str, float]:
         """Returns a dictionary of parameters to execute a circuit at."""

--- a/mitiq/zne/scaling.py
+++ b/mitiq/zne/scaling.py
@@ -15,7 +15,17 @@
 
 """Functions for local and global unitary folding on supported circuits."""
 from copy import deepcopy
-from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple, Union
+from typing import (
+    Any,
+    Callable,
+    Collection,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
 from functools import wraps
 
 import numpy as np
@@ -63,9 +73,7 @@ def _is_measurement(op: ops.Operation) -> bool:
     return isinstance(op.gate, ops.measurement_gate.MeasurementGate)
 
 
-def _pop_measurements(
-    circuit: Circuit,
-) -> List[List[Union[int, ops.Operation]]]:
+def _pop_measurements(circuit: Circuit,) -> List[Tuple[int, ops.Operation]]:
     """Removes all measurements from a circuit.
 
     Args:
@@ -74,15 +82,13 @@ def _pop_measurements(
     Returns:
         measurements: List of measurements in the circuit.
     """
-    measurements = [
-        list(m) for m in circuit.findall_operations(_is_measurement)
-    ]
+    measurements = list(circuit.findall_operations(_is_measurement))
     circuit.batch_remove(measurements)
     return measurements
 
 
 def _append_measurements(
-    circuit: Circuit, measurements: List[Union[int, ops.Operation]]
+    circuit: Circuit, measurements: List[Tuple[int, ops.Operation]]
 ) -> None:
     """Appends all measurements into the final moment of the circuit.
 
@@ -90,11 +96,11 @@ def _append_measurements(
         circuit: a quantum circuit as a :class:`cirq.Circuit`.
         measurements: measurements to perform.
     """
+    new_measurements: List[Tuple[int, ops.Operation]] = []
     for i in range(len(measurements)):
-        measurements[i][0] = (
-            len(circuit) + 1
-        )  # Make sure the moment to insert into is the last in the circuit
-    circuit.batch_insert(measurements)
+        # Make sure the moment to insert into is the last in the circuit
+        new_measurements.append((len(circuit) + 1, measurements[i][1]))
+    circuit.batch_insert(new_measurements)
 
 
 def _check_foldable(circuit: Circuit) -> None:
@@ -150,6 +156,7 @@ def convert_to_mitiq(circuit: QPROGRAM) -> Tuple[Circuit, str]:
         circuit: Mitiq circuit equivalent to input circuit.
         input_circuit_type: Type of input circuit represented by a string.
     """
+    conversion_function: Callable[[QPROGRAM], Circuit]
     if "qiskit" in circuit.__module__:
         from mitiq.mitiq_qiskit.conversions import from_qiskit
 
@@ -284,7 +291,7 @@ def _fold_gates_in_moment(
 def _fold_gates(
     circuit: Circuit,
     moment_indices: Iterable[int],
-    gate_indices: List[Iterable[int]],
+    gate_indices: List[Collection[int]],
 ) -> Circuit:
     """Returns a new circuit with specified gates folded.
 
@@ -359,7 +366,7 @@ def _get_weight_for_gate(
     elif "triple" in weights.keys() and len(op.qubits) == 3:
         weight = weights["triple"]
 
-    if op.gate in _cirq_gates_to_string_keys.keys():
+    if op.gate and op.gate in _cirq_gates_to_string_keys.keys():
         # Get the string key for this gate
         key = _cirq_gates_to_string_keys[op.gate]
         if key in weights.keys():
@@ -473,6 +480,7 @@ def fold_gates_from_left(
 
     # Determine the stopping condition for folding
     ngates = len(list(folded.all_operations()))
+    weights: Optional[Dict[str, float]]
     if fidelities:
         weights = {k: 1.0 - f for k, f in fidelities.items()}
         total_weight = _compute_weight(folded, weights)
@@ -508,6 +516,8 @@ def fold_gates_from_left(
                 if not (kwargs.get("squash_moments") is False):
                     folded = squash_moments(folded)
                 return folded
+
+    return folded
 
 
 @converter
@@ -573,7 +583,7 @@ def fold_gates_from_right(
     circuit = deepcopy(circuit)
     measurements = _pop_measurements(circuit)
 
-    reversed_circuit = Circuit(reversed(circuit))
+    reversed_circuit = Circuit(reversed(circuit.moments))
     reversed_folded_circuit = fold_gates_from_left(
         reversed_circuit,
         scale_factor,
@@ -716,6 +726,7 @@ def fold_gates_at_random(
 
     # Determine the stopping condition for folding
     ngates = len(list(folded.all_operations()))
+    weights: Optional[Dict[str, float]]
     if fidelities:
         weights = {k: 1.0 - f for k, f in fidelities.items()}
         total_weight = _compute_weight(folded, weights)
@@ -785,8 +796,8 @@ def fold_gates_at_random(
 def _fold_local(
     circuit: Circuit,
     scale_factor: float,
-    fold_method: Callable[[Circuit, float, Tuple[Any]], Circuit],
-    fold_method_args: Tuple[Any] = (),
+    fold_method: Callable[..., Circuit],
+    fold_method_args: Optional[Tuple[Any]] = None,
     **kwargs,
 ) -> Circuit:
     """Helper function for implementing a local folding method (which nominally
@@ -834,9 +845,12 @@ def _fold_local(
 
     while scale_factor > 1.0:
         this_stretch = 3.0 if scale_factor > 3.0 else scale_factor
-        folded = fold_method(
-            folded, this_stretch, *fold_method_args, squash_moments=False
-        )
+        if fold_method_args:
+            folded = fold_method(
+                folded, this_stretch, *fold_method_args, squash_moments=False
+            )
+        else:
+            folded = fold_method(folded, this_stretch, squash_moments=False)
         scale_factor /= 3.0
 
     if not (kwargs.get("squash_moments") is False):

--- a/mitiq/zne/zne.py
+++ b/mitiq/zne/zne.py
@@ -97,7 +97,7 @@ def zne_decorator(
     factory: Factory = None,
     scale_noise: Callable[[QPROGRAM, float], QPROGRAM] = fold_gates_at_random,
     num_to_average: int = 1,
-) -> Callable[[QPROGRAM], float]:
+) -> Callable[[Callable[[QPROGRAM], float]], Callable[[QPROGRAM], float]]:
     """Decorator which adds error mitigation to an executor function, i.e., a
     function which executes a quantum circuit with an arbitrary backend and
     returns an expectation value.

--- a/requirements.txt
+++ b/requirements.txt
@@ -11,6 +11,7 @@ pytest~=5.4.1
 pytest-cov~=2.8.1
 flake8~=3.7.9
 black~=19.10b0
+mypy~=0.782
 
 # to run examples
 matplotlib~=3.2.1

